### PR TITLE
[Rllib] Enable tf eager tracing on the training side

### PR DESCRIPTION
**Pull request by kouroshHakha**
## Why are these changes needed?
This enables tf tracing functionality on the training side for algorithms. 

The plot below shows how the new learner API slightly out-performs the old training stack (it has better throughput). It also show that the performance improvements due to enabling tracing on the new learner API stack is on-par with the improvements from enabling tracing on the old stack. 
image

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/learner/tf/tf_learner.py [hunk 1]
```diff
@@ -268,9 +269,24 @@ def helper(_batch):
...
             self.apply_gradients(gradients)
+
+            # NOTE (Kourosh) The reason for returning fwd_out is that it is optionally
+            # needed for compiling the results in a later step (e.g. in
+            # compile_results), but it should not contain anything but tensors, None or
+            # ExtensionTypes, otherwise the tf.function will yell at us because it
+            # won't be able to convert the returned objects to a tensor representation
+            # (for internal reasons). So, in here, we remove anything from fwd_out that
+            # is not a tensor, None or ExtensionType.
+            def filter_fwd_out(x):
```
Comment (avnishn) [inline rllib/core/learner/tf/tf_learner.py:280]: makes sense to me :)

File: rllib/policy/eager_tf_policy.py [hunk 2]
```diff
@@ -174,8 +174,15 @@ def compute_actions_from_input_dict(
         ) -> Tuple[TensorType, List[TensorType], Dict[str, TensorType]]:
             """Traced version of Policy.compute_actions_from_input_dict."""
 
+            # NOTE: In the new RLModule stack the sampling side is not traced with this
```
Comment (avnishn) [inline rllib/policy/eager_tf_policy.py:177]: I think we could trace the policy, and to do that we need to disable autograph. I'll try enabling autograph to confirm.

**Review by avnishn — APPROVED** [review]
LGTM. 

I think I have a solution for making the policies traced. I think that is preventing this is a line where a feature called autograph is being disabled by our code. If we re enable it it could solve our problem.

File: rllib/core/learner/tf/tf_learner.py [hunk 3]
```diff
@@ -268,9 +269,24 @@ def helper(_batch):
...
+                if isinstance(
+                    x, (tf.Tensor, type(None), tf.experimental.ExtensionType)
+                ):
+                    return x
+                return None
+
             return {
                 "loss": loss,
-                "fwd_out": fwd_out,
+                "fwd_out": tree.map_structure(filter_fwd_out, fwd_out),
```
Comment (gjoliver) [inline rllib/core/learner/tf/tf_learner.py:289]: you are replacing the actual output with None.
do you want to actually filter the column out of fwd_out, something like:
```
filter(lambda x: filter_fwd_out is not None, fwd_out)
```

**Review by gjoliver — APPROVED** [review]
looks good. just a nit.

File: rllib/core/learner/tf/tf_learner.py (hunk 3 above)
Comment (kouroshHakha) [reply rllib/core/learner/tf/tf_learner.py:289]: No I actually want it to exist, so that it won't just magically vanish.

**Pull request merged**